patterns/java/src/refactoring_guru/patterns/decorator/example/decorators/CompressionDecorator.java

```diff
@@ -0,0 +1,64 @@
+package refactoring_guru.patterns.decorator.example.decorators;
+
+import java.io.*;
+import java.util.Scanner;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+import java.util.zip.ZipOutputStream;
+
+public class CompressionDecorator extends DataSourceDecorator {
+    FileDataSource fileDataSource;
+
+    public CompressionDecorator(FileDataSource source) {
+        super(source);
+        this.fileDataSource = source;
+    }
+
+    @Override
+    public void writeData(String data) {
+        super.writeData(data);
+        try {
+            File file = new File(fileDataSource.path + fileDataSource.name);
```

Comment:
Same issue applies to the main Application class. It should be aware only of DataSource interface. And it should accept the datasource object from ApplicationConfigurator.

So the single class which knows about specifics should be the ApplicationConfigurator.